repo/blob/gcs/gcs_versioned.go

```diff
@@ -0,0 +1,122 @@
+package gcs
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/pkg/errors"
+	"google.golang.org/api/iterator"
+
+	"github.com/kopia/kopia/repo/blob"
+)
+
+// versionMetadata has metadata for a single BLOB version.
+type versionMetadata struct {
+	blob.Metadata
+
+	// Versioning related information
+	IsDeleteMarker bool
+	Version        string
+}
+
+// versionMetadataCallback is called when processing the metadata for each blob version.
+type versionMetadataCallback func(versionMetadata) error
+
+// IsVersioned returns whether versioning is enabled in the store.
+// It returns true even if versioning is enabled but currently suspended for the
+// bucket. Notice that when object locking is enabled in a bucket, object
+// versioning is enabled and cannot be suspended.
+func (gcs *gcsStorage) IsVersioned(ctx context.Context) (bool, error) {
+	attrs, err := gcs.bucket.Attrs(ctx)
+	if err != nil {
+		return false, errors.Wrapf(err, "could not get versioning info for %s", gcs.BucketName)
+	}
+
+	return attrs.VersioningEnabled, nil
+}
+
+// getBlobVersions lists all the versions for the blob with the given ID.
+func (gcs *gcsStorage) getBlobVersions(ctx context.Context, prefix blob.ID, callback versionMetadataCallback) error {
+	var foundBlobs bool
+
+	if err := gcs.list(ctx, prefix, true, func(vm versionMetadata) error {
+		foundBlobs = true
+
+		fmt.Printf("Found blobs\n")
+		return callback(vm)
+	}); err != nil {
+		return err
+	}
+
+	if !foundBlobs {
+		return blob.ErrBlobNotFound
+	}
+
+	return nil
+}
+
+// listBlobVersions lists all versions for all the blobs with the given blob ID prefix.
+func (gcs *gcsStorage) listBlobVersions(ctx context.Context, prefix blob.ID, callback versionMetadataCallback) error {
```

Comment:
```suggestion
func (gcs *gcsPointInTimeStorage) listBlobVersions(ctx context.Context, prefix blob.ID, callback versionMetadataCallback) error {
```
I guess everything in here should be `gcsPointInTimeStorage` rather than `gcsStorage`